### PR TITLE
GFSv16.3.5[6] GSI updates

### LINKED ISSUE #1321: gdasanal job fails with missing crtm file when testing 2023 date
**Expected behavior**

The gdasanal job should complete successfully and have needed crtm files as its disposal.

**Current behavior**

A fresh high-res C768C384L127 cycled atmos-only test on WCOSS2 started with checksum'd warm-starts from ops 2023021518 fails in the first full cycle gdasanal job with the following error:
```
nid003394.dogwood.wcoss2.ncep.noaa.gov 761:   in setupw, num_bad_ikx ( ikx<1 or ikx>nconvtype ) =        10153
nid002066.dogwood.wcoss2.ncep.noaa.gov 0:            0 setuprad: passive obs          42 ssmis_f18           
nid002066.dogwood.wcoss2.ncep.noaa.gov 0:            0 setuprad: passive obs          59 hirs4_n19           
nid002066.dogwood.wcoss2.ncep.noaa.gov 0:            0 setuprad: passive obs          67 hirs4_metop-b       
nid003278.dogwood.wcoss2.ncep.noaa.gov 566:  SpcCoeff_ReadFile(Binary)(FAILURE) : File ./crtm_coeffs/abi_g18.SpcCoeff.bin not found.
 CRTM_SpcCoeff_Load(FAILURE) : Error reading SpcCoeff file #1, ./crtm_coeffs/abi_g18.SpcCoeff.bin; Process ID: 566
 CRTM_Init(FAILURE) : Error loading SpcCoeff data; Process ID: 566        
 crtm_interface*init_crtm:  ***ERROR*** crtm_init error_status=           3
    TERMINATE PROGRAM EXECUTIONnid003278.dogwood.wcoss2.ncep.noaa.gov 566: MPICH ERROR [Rank 566] [job id 399d5969-f18a-4aa9-a65c-dd52abe7c439] [Fri Feb 17 02:59:08 2023] [nid003278] - Abort(71) (rank 566 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 71) - process 566

aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 71) - process 566
nid003277.dogwood.wcoss2.ncep.noaa.gov 548:  SpcCoeff_ReadFile(Binary)(FAILURE) : File ./crtm_coeffs/abi_g18.SpcCoeff.bin not found.
 CRTM_SpcCoeff_Load(FAILURE) : Error reading SpcCoeff file #1, ./crtm_coeffs/abi_g18.SpcCoeff.bin; Process ID: 548
 CRTM_Init(FAILURE) : Error loading SpcCoeff data; Process ID: 548        
 crtm_interface*init_crtm:  ***ERROR*** crtm_init error_status=           3
    TERMINATE PROGRAM EXECUTIONnid003277.dogwood.wcoss2.ncep.noaa.gov 548: MPICH ERROR [Rank 548] [job id 399d5969-f18a-4aa9-a65c-dd52abe7c439] [Fri Feb 17 02:59:08 2023] [nid003277] - Abort(71) (rank 548 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 71) - process 548
```

Log: /lfs/h2/emc/ptmp/kate.friedman/comrot/deveff2otest/logs/2023021600/gdasanal.log.2

The gfsanal job in the same cycle does not fail. Log: /lfs/h2/emc/ptmp/kate.friedman/comrot/deveff2otest/logs/2023021600/gfsanal.log

Cold-start tests using early 2022 ICs do not fail with this error. Is this error due to the date of ICs being after v16.3.5[6] went into ops? The crtm file appears to be for GOES-18? I also do no see an error related to this crtm file in ops.

The crtm file it croaks on does not exist in either `crtm/2.3.0` (used here) or `crtm/2.4.0`:
```
kate.friedman@dlogin01:/lfs/h2/emc/ptmp/kate.friedman/comrot/deveff2otest> ll /apps/ops/prod/libs/intel/19.1.3.304/crtm/2.3.0/fix/abi_g18.SpcCoeff.bin
ls: cannot access '/apps/ops/prod/libs/intel/19.1.3.304/crtm/2.3.0/fix/abi_g18.SpcCoeff.bin': No such file or directory
kate.friedman@dlogin01:/lfs/h2/emc/ptmp/kate.friedman/comrot/deveff2otest> ll /apps/ops/prod/libs/intel/19.1.3.304/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin                    
ls: cannot access '/apps/ops/prod/libs/intel/19.1.3.304/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin': No such file or directory
```

This crtm file is symlinked into the rundir for both anal jobs in this test (and in other cold-start tests completed for other dates recently) but it's non-existence doesn't appear to cause issues in those other jobs/tests:
```
+ exglobal_atmos_analysis.sh[411]: /bin/ln -sf /apps/ops/prod/libs/intel/19.1.3.304/crtm/2.3.0/fix/abi_g18.SpcCoeff.bin ./crtm_coeffs/abi_g18.SpcCoeff.bin
+ exglobal_atmos_analysis.sh[412]: /bin/ln -sf /apps/ops/prod/libs/intel/19.1.3.304/crtm/2.3.0/fix/abi_g18.TauCoeff.bin ./crtm_coeffs/abi_g18.TauCoeff.bin
```

Thinking I might need recent ops updates to fix files and info files (v16.3.5) I pointed my clone `fix/gsi` folder to the ops `fix/fix_gsi` folder. Same error occurred though (see most recent two gdasanal.log files).

@CoryMartin-NOAA @RussTreadon-NOAA @ADCollard Have you run into a similar error with the analysis for recent dates? Thanks for any help!

**Machines affected**

Only tested on WCOSS2 but likely happens on all machines.

**To Reproduce**

Set up a warm-start test on WCOSS2 with the current `develop` branch. Run checksum on warm-starts for use with updated forecast model.

### LINKED ISSUE #1322: Pull GFSv16.3.5[6] GSI updates into develop
**Description**

Several small updates to the GSI have occurred in operations recently:

1. GFSv16.3.5 - allow continued use of atmospheric motion vectors from the GOES-W satellite as GOES-18 replaces GOES-17 and to include the assimilation of VIIRS radiances from S-NPP and NOAA-20 (which were to be included in the v16.3 implementation but were omitted due to data availability issues). See release notes: https://github.com/NOAA-EMC/global-workflow/blob/dev/gfs.v16/docs/Release_Notes.gfs.v16.3.5.md
2. GFSv16.3.6 - remove GMI from operational processing in the GSI. See release notes: https://github.com/NOAA-EMC/global-workflow/blob/dev/gfs.v16/docs/Release_Notes.gfs.v16.3.6.md

Make workflow-side changes related to this operational upgrades and update GSI hash when GSI-side work is completed. 

- [x] remove `crtm_ver` hack in `build_gsi_enkf.sh` introduced in PR #1319
- [x] remove line in `scripts/exglobal_atmos_analysis.sh` (done at NOAA-EMC/GSI hash [8c3609](https://github.com/NOAA-EMC/GSI/commit/8c360974de29d5e3d1f87d46b8fecb14c4972170).) Will resolve issue #1321.
- [x] disable GMI in `scripts/exglobal_atmos_analysis.sh` (done at https://github.com/NOAA-EMC/GSI/commit/e268acf909e7ea0daaa88b93299168bd654eed99):
```
#GMI1CRBF=${GMI1CRBF:-${COMIN_OBS}/${OPREFIX}gmi1cr.tm00.bufr_d${OSUFFIX}} # GMI temporarily disabled due to array overflow.
#$NLN $GMI1CRBF         gmibufr   # GMI temporarily disabled due to array overflow.
```
- [x] update GSI hash once (https://github.com/NOAA-EMC/GSI/issues/518 & https://github.com/NOAA-EMC/GSI/issues/520 complete) - new hash is [31b8b29](https://github.com/NOAA-EMC/GSI/commit/31b8b29a098fd8df50a9f0b11c6912d47ae0628c).

**Dependencies**

- [x] https://github.com/NOAA-EMC/GSI/issues/518
- [x] https://github.com/NOAA-EMC/GSI/issues/520

### PR DESCRIPTION
**Pull request by KateFriedman-NOAA**
**Description**

This PR brings in GFSv16.3.5[6] GSI updates into `develop`:

- New GSI hash [31b8b29](https://github.com/NOAA-EMC/GSI/commit/31b8b29a098fd8df50a9f0b11c6912d47ae0628c).
- Workflow updates to match:
  - remove `crtm_ver` hack in `build_gsi_enkf.sh` introduced in PR https://github.com/NOAA-EMC/global-workflow/pull/1319
  - remove line in `scripts/exglobal_atmos_analysis.sh` (done at NOAA-EMC/GSI hash [8c3609](https://github.com/NOAA-EMC/GSI/commit/8c360974de29d5e3d1f87d46b8fecb14c4972170).) Will resolve issue https://github.com/NOAA-EMC/global-workflow/issues/1321.
  - disable GMI in `scripts/exglobal_atmos_analysis.sh` (done at https://github.com/NOAA-EMC/GSI/commit/e268acf909e7ea0daaa88b93299168bd654eed99)

Resolves #1321
Resolves #1322

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

- [x] Clone and build tests on Hera
- [x] Cycled test on Hera (5.5 cycles C96C48L127)
- [x] Tested pieces previously in high-res real-time tests on WCOSS2.
- [x] Changes running in GFSv16 operations currently.
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

### CODE REVIEW

File: Externals.cfg
Comment (RussTreadon-NOAA) [inline Externals.cfg:32]: This hash is fine but FYI, NOAA-EMC/GSI develop has new commits.  The current head is [`89ca542`](https://github.com/NOAA-EMC/GSI/commit/89ca542c963c7887494e728656515918ce7a8518).  Five open PRs will be likely merged into develop over the coming weeks.

File: scripts/exglobal_atmos_analysis.sh
Comment (RussTreadon-NOAA) [inline scripts/exglobal_atmos_analysis.sh:469]: @emilyhcliu , @ADCollard , @HaixiaLiu-NOAA, @xincjin-NOAA.   What GMI behavior do you want from the `develop` g-w?  Retain, skip, or something else?

File: scripts/exglobal_atmos_analysis.sh
Comment (RussTreadon-NOAA) [inline scripts/exglobal_atmos_analysis.sh:869]: @emilyhcliu , @ADCollard , @HaixiaLiu-NOAA.  What ABI behavior do you want from the `develop` g-w?  Retain, skip, or something else?

File: Externals.cfg
Comment (KateFriedman-NOAA) [reply Externals.cfg:32]: @RussTreadon-NOAA Noted, thanks! I'll let those other incoming PRs update the GSI hash further.

File: sorc/checkout.sh
Comment (RussTreadon-NOAA) [inline sorc/checkout.sh:164]: Same GSI `develop` hash comment as for `Externals.cfg`

File: scripts/exglobal_atmos_analysis.sh
Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:469]: Thanks for checking on this with other DA folks @RussTreadon-NOAA !

File: scripts/exglobal_atmos_analysis.sh
Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: Thanks for checking on this with other DA folks @RussTreadon-NOAA !

Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: Note, when I ran with this line included for a more recent date in a high-res test on WCOSS2 it failed. See issue #1321.

File: sorc/build_gsi_enkf.sh
Comment (RussTreadon-NOAA) [inline sorc/build_gsi_enkf.sh:23]: Just curious - why were we temporarily setting `crtm_ver=2.4.0` here?

**Review by RussTreadon-NOAA — COMMENTED** [review]
Changes look OK.   Note that GSI hash not at current head.    Questions for DA developers as to the preferred g-w behavior for ABI and GMI.

File: scripts/exglobal_atmos_analysis.sh
Comment (RussTreadon-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: Thanks @KateFriedman-NOAA  for noting this.  The failure is due to problems with the crtm coefficients.   Ideally the crtm team provides a fix instead of g-w sidestepping the problem.

File: sorc/build_gsi_enkf.sh
Comment (KateFriedman-NOAA) [reply sorc/build_gsi_enkf.sh:23]: The GSI hash we were using previously was building with `crtm/2.3.0`. I forced it to build with `crtm/2.4.0` with this setting as part of updating the whole global-workflow to `crtm/2.4.0` in PR #1319, to avoid waiting for the GSI PRs that just went in to go in. We don't yet use version files in `develop` so this was how I did it.

File: scripts/exglobal_atmos_analysis.sh
Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: > Ideally the crtm team provides a fix instead of g-w sidestepping the problem.

That would be ideal...who do we speak with to get such a fix in place?

File: sorc/build_gsi_enkf.sh
Comment (RussTreadon-NOAA) [reply sorc/build_gsi_enkf.sh:23]: Thank you, @KateFriedman-NOAA , for the explanation.

File: scripts/exglobal_atmos_analysis.sh
Comment (RussTreadon-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: Ben Johnson comes to mind but I'm not sure if he is the right person to contact.   @ADCollard , @emilyhcliu , @HaixiaLiu-NOAA can probably point us at the right person.   Once we have the right coefficients, the library team needs to add the corrected files to `CRTM_FIX` on various machines.   Doing so on WCOSS2 will likely require an increment in the CRTM version number (NCO policy).   We need to decide if such a change is worth the ripples it creates.

Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: @RussTreadon-NOAA Gotcha, thanks for those details! Given the fact that the `crtm` version will likely increment to accommodate this fix we can plan to leave this line out while still using `crtm/2.4.0` and add it back in should we get a new `crtm/2.5.0.` Please advise otherwise though, thanks!

Comment (RussTreadon-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: The change is required given problems with the crtm coefficients.   We can't wait for the crtrm files to be fixed.   The danger in removing functionality is that (1) the reason for the removal may be forgotten, (2) a fix is not actively pursued.

**Comment by KateFriedman-NOAA** [comment]
@emilyhcliu @ADCollard @HaixiaLiu-NOAA @xincjin-NOAA Please see the questions above from @RussTreadon-NOAA . Need feedback from you before we merge this PR. Thanks!

File: scripts/exglobal_atmos_analysis.sh
Comment (xincjin-NOAA) [reply scripts/exglobal_atmos_analysis.sh:469]: Sorry for the later response, I thought other persons might response. I would suggest to skip it for now, because it does not work properly.

File: scripts/exglobal_atmos_analysis.sh
Comment (HaixiaLiu-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: > @emilyhcliu , @ADCollard , @HaixiaLiu-NOAA. What ABI behavior do you want from the `develop` g-w? Retain, skip, or something else?

The abi_g18 coeff files are in the crtmv2.4.0 installation already, so we can keep the abi in the develop g-w now. Thanks.

File: scripts/exglobal_atmos_analysis.sh
Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:469]: Thanks @xincjin-NOAA for the feedback. Based on your reply I will leave this line commented out as it is in this PR (e.g. skip it). We can uncomment it in the future should it be fixed.

File: scripts/exglobal_atmos_analysis.sh
Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: @HaixiaLiu-NOAA The `abi_g18.SpcCoeff.bin` file is not in the WCOSS2 production `crtm/2.4.0` install, nor the hpc-stack `crtm/2.4.0` installs. Hence this related global-workflow issue and why I deleted this line, see https://github.com/NOAA-EMC/global-workflow/issues/1321.

```
kate.friedman@dlogin01:~> ll /apps/ops/prod/libs/intel/19.1.3.304/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin
ls: cannot access '/apps/ops/prod/libs/intel/19.1.3.304/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin': No such file or directory
```
```
-bash-4.2$ ll /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-18.0.5.274/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin
ls: cannot access /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-18.0.5.274/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin: No such file or directory
-bash-4.2$ ll /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-2022.1.2/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin
ls: cannot access /scratch2/NCEPDEV/nwprod/hpc-stack/libs/hpc-stack/intel-2022.1.2/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin: No such file or directory
```

I do see it in the newer EPIC hpc-stack `crtm/2.4.0` installs though (good):
```
-bash-4.2$ ll /scratch1/NCEPDEV/nems/role.epic/hpc-stack/libs/intel-2022.1.2/intel-2022.1.2/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin
-rw-r--r-- 1 role.epic nems 872 Nov 10 19:50 /scratch1/NCEPDEV/nems/role.epic/hpc-stack/libs/intel-2022.1.2/intel-2022.1.2/crtm/2.4.0/fix/abi_g18.SpcCoeff.bin
```
...but we aren't yet using those stacks (I am currently testing them on Hera/Orion).

To be able to add this line back in we'd need:

1. The WCOSS2 `crtm/2.4.0` updated to add in the missing `abi_g18.SpcCoeff.bin` file.
2. Finish moving to the EPIC hpc-stacks elsewhere.

The EPIC hpc-stacks are intel 2022 but the GSI is still using intel 2018. I can force the workflow to load the EPIC hpc-stack intel 2022 `crtm/2.4.0` at runtime but the GSI would still be building with the pre-EPIC hpc-stack intel 2018 `crtm/2.4.0`. It should work but will continue the inconsistency between the workflow and GSI. The GSI needs to be updated to intel 2022 regardless (another issue).

As it stands right now, the `develop` branch will break for newer CDATEs with this line included and us using the `crtm/2.4.0` fix files as they are now. As observed in https://github.com/NOAA-EMC/global-workflow/issues/1321.

Please advise. Perhaps I can comment out this line for now with a note to uncomment when we move to `crtm/2.4.0` that includes the `abi_g18.SpcCoeff.bin` file everywhere? Or is there another way to temporarily disable this without removing the line? Thanks!

**Comment by emilyhcliu** [comment]
Good morning,

For some reason, I could not leave my comment in the [PR](https://github.com/NOAA-EMC/global-workflow/pull/1404) earlier this morning.
Here is my comment:

ObsProc team had identified the GMI problem (filesize doubled) a while ago.  The GMI data was dumped twice.  
Currently, a corn job is set to process GMI (from March 9, 2023) and stored in the following local directory:
 /lfs/h2/emc/obsproc/noscrub/iliana.genkova/CRON/FIXGMI/com/obsproc/v1.1/

This is temporary.  For g-w, I suggest keeping it as is: commenting out GMI data.  
For testing, we increased the size of the array which holds the GMI data in the read routine while testing, so it is not a problem reading the GMI (doubled filesize) from the global dump.  

Emily

**Comment by KateFriedman-NOAA** [comment]
> This is temporary. For g-w, I suggest keeping it as is: commenting out GMI data.

@emilyhcliu Thanks for your feedback and update on the resolution of this issue for operations.

Is there a timeline for a fix getting into operations for both the size of the GMI data and/or uncommenting the line in the GSI script? Will the size of the array for the GMI be changed to be double so future runs using the affected dump files can process them?

I will leave this line commented out in global-workflow `develop` for now.

**Comment by emilyhcliu** [comment]
> > This is temporary. For g-w, I suggest keeping it as is: commenting out GMI data.
> 
> @emilyhcliu Thanks for your feedback and update on the resolution of this issue for operations.
> 
> Is there a timeline for a fix getting into operations for both the size of the GMI data and/or uncommenting the line in the GSI script? Will the size of the array for the GMI be changed to be double so future runs using the affected dump files can process them?
> 
> I will leave this line commented out in global-workflow `develop` for now.

In GSI develop, we will not double the size of the GMI array in the read routine.  We probably will add a check to skip the duplicated data.

I will check with Iliana for the timeline.  The possible time to add the correct data processing for GMI  is when we do the N21 upgrade.

File: scripts/exglobal_atmos_analysis.sh
Comment (HaixiaLiu-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: @KateFriedman-NOAA You are correct. I checked the Hera EPIC hpc-stack installation and noticed that the abi_g18 coeff files exist there. The WCOSS2 crtmv2.4.0 installation needs an update to include N21 and G18 coeff files and is awaiting Andrew's approval. Andrew is currently testing the N21 coeff files and will let Hang Lei know if the N21 coeff files are good to go this week, at which point the WCOSS2 v2.4.0 can be updated to include these new coeff files.

Your suggestion of commenting out the abi_g18 for now sounds good to me.

Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: @HaixiaLiu-NOAA Sounds good, thanks for confirming we'll be able to uncomment once we ingest the upcoming ops updates related to N21. When I process the upcoming GSI ops update in global-workflow I will replicate the changes into the `develop` branch as well. I'll leave the abi_g18 line commented out for now.

Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: > I'll leave the abi_g18 line commented out for now.

@HaixiaLiu-NOAA Slight correction to my reply...I will leave the abi_g18 _removed_ (not commented out) for now, which mimics the change made in ops recently. I will add the line back in when it's added back in ops.

File: scripts/exglobal_atmos_analysis.sh
Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:469]: Added note in https://github.com/NOAA-EMC/global-workflow/issues/1356 to uncomment line when ingesting upcoming ops updates.

File: scripts/exglobal_atmos_analysis.sh
Comment (KateFriedman-NOAA) [reply scripts/exglobal_atmos_analysis.sh:869]: Added note to https://github.com/NOAA-EMC/global-workflow/issues/1356 to add line back in when processing upcoming ops updates.

**Comment by KateFriedman-NOAA** [comment]
@emilyhcliu @ADCollard @HaixiaLiu-NOAA @xincjin-NOAA Thanks for your feedback regarding the abi_g18 and GMI lines in `scripts/exglobal_atmos_analysis.sh`. Based on feedback I am retaining the changes in this PR and have made notes to revert in future PRs (see #1356).

I believe all reviewer comments have been addressed. This PR is ready for final reviews. @WalterKolczynski-NOAA @aerorahul

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Pull request merged**
